### ISSUE TITLE
Collision not working as expected

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#10071]
##### Minetest version

```
5.2.0
5.3-dev
```

##### Summary

Collision should not prevent one object from passing through or leaving another. See the following illustration:

![Illustration](https://user-images.githubusercontent.com/34514239/85184033-81bd7a00-b28e-11ea-8b9d-dd799dadd76c.png)

The left face of one object should only collide with the right face of other objects, the top face should only collide with the bottom face of other objects etc... Currently, it seems that left faces are colliding with left faces etc.

**Comment by TheTermos** [comment]
Collision detection in 5.3 is quite different than that of 5.2, one thing in common is what I call clipping recovery. What it does is boxes can collide even when already intersecting - to a certain degree.

I don't know why it was introduced originally, but it's been implemented in 5.3 because:
- it was there before, so if it were absent mechanics that grew to rely on it would behave differently and that would have been perceived as bugs.
- the fatal flaw in MT client server design giving authority over player movement to the client, causing players to be constantly out of sync with entities and each other. One side effect is that collision detection involving moving players and other moving players/entities nearly always fails (see also #9806).
- another fundamental engine flaw: 32 bit precision of object/player position, there are simply too few precise digits to work with and collision detection is the more likely to fail the farther from the world center.

However exact left-to-left collision you describe isn't possible in 5.3. It is possible in 5.2 but only for boxes smaller that 0.27.


That said, your problem description suggests you are trying to achieve something relying on boxes intersecting, and that's usually not a great idea. If that's the case please describe what you're trying to do, there are usually better ways to do it.

**Comment by appgurueu** [comment]
> That said, your problem description suggests you are trying to achieve something relying on boxes intersecting, and that's usually not a great idea. If that's the case please describe what you're trying to do, there are usually better ways to do it.

You are right. I am in fact using a "better way" now - my use case is a "snake" game. For smooth animations, the tail is entering the last segment while the head is coming out of the first. As my segments used to be entities and I require them to be physical (for snake-snake collisions), it didn't work. So I went with nodes instead of entities for the segments, which allows me to use "tunnel"-collisionboxes, also increasing performance.

**Comment by paramat** [comment]
> Collision should not prevent one object from passing through or leaving another [...]
The left face of one object should only collide with the right face of other objects, the top face should only collide with the bottom face of other objects etc

That is a matter of opinion and depends on the situation and your needs, it is not a bug.
It makes sense to me that if 2 objects are intersecting a collision occurrs.

**Comment by ResuUman** [comment]
Don't know what appgurueu meant but... Please consider following situation:
While swimming on the boat you are seated but your head is over 1m (pos: 1.6). 
It seams that collision detection with (eg.?) stone is done with boat only and not with head.
In result one can swim under block (0-1m) with head is inside the stone block (1-2m). And you are blinded. And you are confused with black screen.

**Comment by paramat** [comment]
That seems unrelated and due to collision being considered for parent object only.
